cmake/modules/BashCompletion.cmake

```diff
@@ -1,93 +1,63 @@
-			PKG_GET_VARIABLE(BASHCOMP_PKG_PATH bash-completion completionsdir)
-		ELSE()
-			# Second, use cmake (preferred but less common)
-			FIND_PACKAGE(bash-completion QUIET)
-			IF(BASH_COMPLETION_FOUND)
-				SET(BASHCOMP_PKG_PATH "${BASH_COMPLETION_COMPLETIONSDIR}")
-			ENDIF()
-		ENDIF()
+# Honor manual override if provided
+if(NOT BASHCOMP_PKG_PATH)
+	# First, use pkg-config, which is the most reliable
+	find_package(PkgConfig QUIET)
+	if(PKGCONFIG_FOUND)
+		PKG_CHECK_MODULES(BASH_COMPLETION bash-completion)
+		PKG_GET_VARIABLE(BASHCOMP_PKG_PATH bash-completion completionsdir)
+	else()
+		# Second, use cmake (preferred but less common)
+		find_package(bash-completion QUIET)
+		if(BASH_COMPLETION_FOUND)
+			set(BASHCOMP_PKG_PATH "${BASH_COMPLETION_COMPLETIONSDIR}")
+		endif()
+	endif()
 
-		# Third, use a hard-coded fallback value
-		IF("${BASHCOMP_PKG_PATH}" STREQUAL "")
-			SET(BASHCOMP_PKG_PATH "/usr/share/bash-completion/completions")
-		ENDIF()
-	ENDIF()
+	# Third, use a hard-coded fallback value
+	if("${BASHCOMP_PKG_PATH}" STREQUAL "")
+		set(BASHCOMP_PKG_PATH "/usr/share/bash-completion/completions")
+	endif()
+endif()
 
-	# Always provide a fallback for non-root INSTALL()
-	SET(BASHCOMP_USER_PATH "${CMAKE_INSTALL_PREFIX}/share/bash-completion/completions")
+# Always provide a fallback for non-root INSTALL()
+set(BASHCOMP_USER_PATH "share/bash-completion/completions")
 
-	# Cmake doesn't allow easy use of conditional logic at INSTALL() time
-	# this is a problem because ${BASHCOMP_PKG_PATH} may not be writable and we
-	# need sane fallback behavior for bundled INSTALL() (e.g. .AppImage, etc).
-	#
-	# The reason this can't be detected by cmake is that it's fairly common to
-	# run "cmake" as a one user (i.e. non-root) and "make install" as another user
-	# (i.e. root).
-	#
-	# - Creates a script called "install_${SCRIPT_NAME}_completion.sh" into the
-	#   working binary directory and invokes this script at install.
-	# - Script handles INSTALL()-time conditional logic for sane ballback behavior
-	#   when ${BASHCOMP_PKG_PATH} is unwritable (i.e. non-root); Something cmake
-	#   can't handle on its own at INSTALL() time)
-	MACRO(BASHCOMP_INSTALL SCRIPT_NAME)
-		# A shell script for wrapping conditionl logic
-		SET(BASHCOMP_SCRIPT "${CMAKE_CURRENT_BINARY_DIR}/install_${SCRIPT_NAME}_completion.sh")
+macro(BASHCOMP_INSTALL SCRIPT_NAME)
+	# Note:  When running from CPack, message(...) will be supressed unless WARNING
+	if(WIN32)
+		message(STATUS "Bash completion is not supported on this platform.")
+	elseif(APPLE)
+		message(STATUS "Bash completion is not yet implemented for this platform.")
+	else()
+		install(FILES "${SCRIPT_NAME}" DESTINATION "${BASHCOMP_USER_PATH}")
+		if(BASHCOMP_PKG_PATH)
+			# TODO: CMake 3.21 Use "file(COPY_FILE ...)"
+			install(CODE "
+					execute_process(COMMAND ${CMAKE_COMMAND} -E copy \"${SCRIPT_NAME}\" \"${BASHCOMP_PKG_PATH}\" ERROR_QUIET RESULT_VARIABLE result)
```

Comment:
Why does it install `SCRIPT_NAME` twice, once in line 47 and once in line 51?
Also, where is the `BASHCOMP_PKG_PATH` variable from?

Reply:
> Why does it install `SCRIPT_NAME` twice, once in line 47 and once in line 51?
It does not install `SCRIPT_NAME` twice.  It installs a and then runs the script 

It's due to cmake's lack of conditional `INSTALL(...)` logic.  For some reason, I removed the comment that describes this.  It's here: https://github.com/LMMS/lmms/pull/4604/commits/00d4dddb6d71eb54279b5b556ef2b566fdd3eba8#diff-7960a523ea4d93951917fb0445e1145899f31c5a796e123ee06c805095e37221R59-R71.

In short:
1. Always install to user-space
2. Attempt to install system-wide, but don't fail the build if it doesn't work.

We did have some talks about this "system wide" install previously because this is the first time that we've had a component to install in a prefix that's outside of our own install prefix.  I suppose we could instead detect a system-wide prefix and handle this in a more standard fashion.  If that's the case, we can fix it as part of this PR if needed.

I think there's a typo in the status message though.... I think this should say `BASHCOMP_PKG_PATH` not `BASHCOMP_USER_PATH`.
```diff
-						message(STATUS  "Unable to install bash-completion support system-wide: ${BASHCOMP_USER_PATH}/${SCRIPT_NAME})
+ -						message(STATUS  "Unable to install bash-completion support system-wide: ${BASHCOMP_PKG_PATH}/${SCRIPT_NAME})
```

> Also, where is the `BASHCOMP_PKG_PATH` variable from?

It's from #4604.  This is not a newly introduced variable with this PR.  I believe its intended to be the `completionsdir` component, found by pkg-config.  My memory is foggy from back then and the conversations seem to have largely occurred on Discord so decisions are not all obvious.  Anyway, we also re-use this variable name for cmake's `find_package(bash-completion)` portion as well.  It's probably a bad name, but I do not plan on changing it as part of this PR but can if needed.

Reply:
> I believe its intended to be the completionsdir component, found by pkg-config.

Indeed! The top of the file actually explains it. You can even set this variable using `-DBASHCOMP_PKG_PATH`

> Furthermore, I think there's a typo in the status message though.... I think this should say BASHCOMP_PKG_PATH not BASHCOMP_USER_PATH.

Agreeing. Feel free to fix.

> Attempt to install system-wide, but don't fail the build if it doesn't work.

IMO, the natural behavior would be: If the user does a local install, install bashcomp locally. If they do a system install ("root"), install it system-wide. I assume that back then, we thought: "An install should always try to be system-wide, because local installed bashcomp is not sourced by your bashrc". However, you could give the same argument for the `PATH` variable (a local install of `bin/lmms` will never be in your `PATH` by default). In both cases, you can extend your bashrc to source a local bashcomp or to extend your `PATH` to a local directory. All in all, I would favor to do the install exactly one, and always in the normal install tree. However, it is just my basic opinion, and also it might be considered out of scope of this PR. Tagging @PhysSong because they were also active in that bashcomp PR in 2018.

Reply:
Thanks for the feedback.  I'll put some time into fixing this.